source/openqasm/ast_visitor/ast_visitor.py

```diff
@@ -0,0 +1,73 @@
+from typing import Optional, TypeVar, Generic
+
+from openqasm.ast import OpenNode
+
+T = TypeVar("T")
+
+
+class NodeVisitor(Generic[T]):
+    """
+    A node visitor base class that walks the abstract syntax tree and calls a
+    visitor function for every node found.  This function may return a value
+    which is forwarded by the `visit` method.
+
+    Modified from the implementation in ast.py in the Python standard library.
+    We added the context argument to the visit method. It allows the visitor
+    to hold temporary state while visiting the nodes.
+
+    The optional context argument in visit/generic_visit methods can be used to hold temporary
+    information that we do not want to hold in either the AST or the visitor themselves.
+    """
+
+    def visit(self, node: OpenNode, context: Optional[T] = None):
+        """Visit a node."""
+        method = "visit_" + node.__class__.__name__
+        visitor = getattr(self, method, self.generic_visit)
+        # The visitor method may not have the context argument.
+        if context:
+            return visitor(node, context)
+        else:
+            return visitor(node)
```

Comment:
If all `visit_` methods take `context: Optional[T] = None`, then you can just `return visitor(node, context)` without branching - the `None` will safely get passed down in the same manner anyway.

(This comment applies to a few places in this file.)

Reply:
One reason that this branch existed is that it gives users not to have the `context` parameter in visit_XXX methods if they do not need it. Once we add some concrete classes of the ast_visitor, it will much more clear.